### ISSUE TITLE
Compatibility

### ISSUE THREAD

**Issue by Coder-256** [beardedspice/beardedspice#668]
I was looking into improving browser support in #96, then I realized that there were a few problems (both with browser support and BeardedSpice itself). I ran into enough problems that I felt that a new issue should be opened.

There are two main ways to communicate with Chromium-based browsers (and Safari*) from native apps:

1. Use the Applescript scripting bridge (SBApplication, etc). This is what BS currently does. However, this has many problems such as lack of support (for example, Vivaldi and Opera, and Safari without enabling insecure Developer modes) and problems with the scripting bridge itself (which, as I am sure you know, is relatively broken and requires many hacky workarounds).
2. Create a HTTP server running on localhost, and have a browser extension connect to that webserver. However, this is very unreliable and buggy, and _extremely_ insecure. I would strongly recommend against this option.
3. This option is by far the best one. As I mentioned earlier, I could create a Chromium extension (something I have done before) that uses Native Messaging to communicate with BeardedSpice. This would be the most stable, secure, and seamless option for the user. However, it is probably the most difficult for the developer.

*Safari would require an [App Extension](https://developer.apple.com/library/content/documentation/NetworkingInternetWeb/Conceptual/SafariAppExtension_PG/index.html), which is much simpler and can be implemented separately.

Options two and three would work for almost all Chromium-based browsers such as Chromium itself, Chrome, Opera, Vivaldi, Yandex Browser, CentBrowser, Slimjet, and many many more. But because option 2 is insecure, I will talk about one possible way of doing option three.

In order to use native messaging, the steps would look something like this:

1. BeardedSpice installs a Native Messaging Host manifest file (just a simple JSON file). However, it must be generated at BS launch and regenerated after BS is moved because it must include the path to `BeardedSpice.app`.
2. That JSON file must be saved in a two (or possibly more) different locations:
  - `~/Library/Application Support/Google/Chrome/NativeMessagingHosts/` (for most browsers)
  - `~/Library/Application Support/Chromium/NativeMessagingHosts/` (for Chromium)
And it must be edited once BS launches if it is in a different path from when it was last written (the file probably should be written on every launch).
3. Users should be prompted to install the extension in whichever browsers they use by opening a link to the Chrome Web Store (or it can be installed directly from the website, ex. [beardedspice.github.io](https://beardedspice.github.io/)). Note that [publishing the extension](https://developer.chrome.com/webstore/publish) is a bit of a difficult process that must be done before the extension can be installed for end users.
4. Each browser with the extension will attempt at a few events (extension install, app relaunch, etc.) to connect to the Native Messaging Host. The host will be an executable embedded within the BeardedSpice app bundle which will handle communication with the extension. The main app will communicate with the host somehow (probably through a `NSPort`), the host will communicate with the extension, and the extension will communicate with the browser. Safari would do this differently, and would be much easier to implement.

This, although it may seem a bit complex, is not too difficult a task. However, there is one huge problem: it would require an enormous refactoring of BS. Like a really really big refactoring. Not just renaming. Lots of code would need to be moved around, rewritten, and in the end, there would be more new code than original code. This is the case for both the Chromium and Safari extensions.

For example, the `TabAdapter` class should be refactored so that refreshing tabs is actually in each implementation instead of in the far-too-large AppDelegate, `versions.plist` should be removed and versions should be read from the `.js` files themselves, `BSCService` **really** needs to stop relying on deprecated APIs.

I'm not exactly sure what to do, and you could reasonably call me a bit crazy for saying this, but I think a full rewrite from the ground up is the best thing to do. Start an entirely new Xcode project, maybe even in Swift, set a high deployment target, and write the code again. But this time forget your previous knowledge and learn the updated way, and there will not be any of the problems I mentioned in the last paragraph. That's what I would suggest, because I think it would make BeardedSpice (an app that does have real potential) into a modern, lightweight app.

This is just a topic for discussion. Everything would still probably work for another year or so as is, but this *will* need to happen eventually, as some of the deprecated functions that BS uses on will likely removed at some point. What do you all think?

**Comment by Coder-256** [comment]
@Stillness-2 @1ps0 Any thoughts yet?

**Comment by 1ps0** [comment]
@Coder-256 That's a pretty good breakdown of what would be needed to expand the app coverage of BeardedSpice. The problem is that it would require a fulltime team and quite a lot of invested work for only a moderate payoff. Maintenance alone would become hell and maintenance is the name of the game for an app at this stage.

I've made several attempts at breaking down AppDelegate and refactoring the codebase. You can see it here: https://github.com/1ps0/beardedspice/tree/chore/optimize-lookup 
Unfortunately it has very little payoff, other than increasing my understanding of how the entire application works. The version in that branch doesn't contain all features either (i'd consider it a quasi-alpha).

I have a completely different approach in mind now, and will detail it later. However, these days I have very little time to perform maintainer's duties much less refactor the large amount of technical debt we're currently sporting.

**Comment by Ashtonian** [comment]
So this is old, would the code still require refactor? 

Would this enable the app to automatically work with an active tab if it was open so the user wouldnt have to pick it?
